File: services/subtraction-service/src/main/java/com/calculator/subtraction/infrastructure/InMemoryCalculationCache.java
package com.calculator.subtraction.infrastructure;

import com.calculator.subtraction.domain.CalculationCache;
import com.calculator.subtraction.domain.CalculationRequest;
import com.calculator.subtraction.domain.CalculationResult;
import org.springframework.cache.annotation.Cacheable;
import org.springframework.cache.annotation.CachePut;
import org.springframework.stereotype.Component;

import java.util.concurrent.ConcurrentHashMap;

@Component
public class InMemoryCalculationCache implements CalculationCache {
    
    private final ConcurrentHashMap<String, CalculationResult> cache = new ConcurrentHashMap<>();
    
    @Override
    @Cacheable("calculations")
    public CalculationResult get(CalculationRequest request) {
        String key = generateKey(request);
        return cache.get(key);
    }
    
    @Override
    @CachePut("calculations")
    public void store(CalculationRequest request, CalculationResult result) {
        String key = generateKey(request);
        cache.put(key, result);
    }
    
    private String generateKey(CalculationRequest request) {
        return "subtract:" + request.operand1() + ":" + request.operand2();
    }
}

File: services/addition-service/src/main/kotlin/com/calculator/addition/infrastructure/InMemoryCalculationCache.kt
package com.calculator.addition.infrastructure

import com.calculator.addition.domain.CalculationCache
import com.calculator.addition.domain.CalculationRequest
import com.calculator.addition.domain.CalculationResult
import org.springframework.cache.annotation.Cacheable
import org.springframework.cache.annotation.CachePut
import org.springframework.stereotype.Component
import java.util.concurrent.ConcurrentHashMap

@Component
class InMemoryCalculationCache : CalculationCache {
    
    private val cache = ConcurrentHashMap<String, CalculationResult>()
    
    @Cacheable("calculations")
    override fun get(request: CalculationRequest): CalculationResult? {
        val key = generateKey(request)
        return cache[key]
    }
    
    @CachePut("calculations")
    override fun store(request: CalculationRequest, result: CalculationResult) {
        val key = generateKey(request)
        cache[key] = result
    }
    
    private fun generateKey(request: CalculationRequest): String {
        return "add:${request.operand1}:${request.operand2}"
    }
}